### PR TITLE
Add SipHash algorithm

### PR DESCRIPTION
**Pull request by SparkiDev**

### CODE REVIEW

File: wolfcrypt/src/siphash.c [hunk 1]
```diff
@@ -0,0 +1,924 @@
...
+                        (a)[7] = (byte)((n) >> 56)
+#endif
+
+/**
+ * Initialize SipHash operation with a key.
+ *
+ * @param [out] sipHash  SipHash object.
+ * @param [in]  key      16 byte array - little endian.
+ * @return  BAD_FUNC_ARG when sipHash or key is NULL.
+ * @return  BAD_FNUC_ARG when outSz is neither 8 nor 16.
```
Comment (dgarske) [inline wolfcrypt/src/siphash.c:145]: Typo: `BAD_FNUC_ARG` -> `BAD_FUNC_ARG`. Plus about 10 more same below.

File: wolfcrypt/src/siphash.c [hunk 2]
```diff
@@ -0,0 +1,924 @@
...
+ *
+ * wolfSSL is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
```
Comment (dgarske) [inline wolfcrypt/src/siphash.c:20]: Can you add a comment at top that assembly optimized versions are supported with GNUC with `__aarch64__` and `__x86_64__`.

File: wolfssl/wolfcrypt/siphash.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+WOLFSSL_API int wc_SipHash(const unsigned char* key, const unsigned char* in,
+    word32 inSz, unsigned char* out, unsigned char outSz);
+
+#ifdef __cplusplus
+    } /* extern "C" */
+#endif
+
+
+#endif /* NO_AES && WOLFSSL_CMAC */
+#endif /* WOLF_CRYPT_CMAC_H */
```
Comment (dgarske) [inline wolfssl/wolfcrypt/siphash.h:92]: Should be `WOLF_CRYPT_SIPHASH_H`

File: wolfssl/wolfcrypt/siphash.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    unsigned char outSz);
+WOLFSSL_API int wc_SipHash(const unsigned char* key, const unsigned char* in,
+    word32 inSz, unsigned char* out, unsigned char outSz);
+
+#ifdef __cplusplus
+    } /* extern "C" */
+#endif
+
+
+#endif /* NO_AES && WOLFSSL_CMAC */
```
Comment (dgarske) [inline wolfssl/wolfcrypt/siphash.h:91]: Should be `WOLFSSL_SIPHASH`

File: wolfssl/wolfcrypt/siphash.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+
+#ifndef WOLF_CRYPT_SIPHASH_H
+#define WOLF_CRYPT_SIPHASH_H
+
+#if defined(WOLFSSL_SIPHASH)
```
Comment (dgarske) [inline wolfssl/wolfcrypt/siphash.h:26]: Please make sure `#include <wolfssl/wolfcrypt/types.h>` is included here above this. Will ensure settings.h and visibility.h are included.... For users who are not using ./configure.

File: wolfcrypt/test/test.c [hunk 6]
```diff
@@ -30334,6 +30345,316 @@ WOLFSSL_TEST_SUBROUTINE int cmac_test(void)
...
 #endif /* NO_AES && WOLFSSL_CMAC */
 
+#if defined(WOLFSSL_SIPHASH)
+
+WOLFSSL_TEST_SUBROUTINE int siphash_test(void)
+{
+    int ret = 0;
+    int i;
+#if WOLFSSL_SIPHASH_CROUNDS == 2 && WOLFSSL_SIPHASH_DROUNDS == 4
+    unsigned char k[16] = { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
```
Comment (dgarske) [inline wolfcrypt/test/test.c:30355]: Can any of these be `const`?

File: wolfcrypt/test/test.c [hunk 7]
```diff
@@ -30334,6 +30345,316 @@ WOLFSSL_TEST_SUBROUTINE int cmac_test(void)
...
+        { 0xe3, 0x04, 0x0c, 0x00, 0xeb, 0x28, 0xf1, 0x53, 
+          0x66, 0xca, 0x73, 0xcb, 0xd8, 0x72, 0xe7, 0x40, },
+        { 0x76, 0x97, 0x00, 0x9a, 0x6a, 0x83, 0x1d, 0xfe, 
+          0xcc, 0xa9, 0x1c, 0x59, 0x93, 0x67, 0x0f, 0x7a, },
+        { 0x58, 0x53, 0x54, 0x23, 0x21, 0xf5, 0x67, 0xa0, 
+          0x05, 0xd5, 0x47, 0xa4, 0xf0, 0x47, 0x59, 0xbd, },
+        { 0x51, 0x50, 0xd1, 0x77, 0x2f, 0x50, 0x83, 0x4a, 
+          0x50, 0x3e, 0x06, 0x9a, 0x97, 0x3f, 0xbd, 0x7c, },
+    };
+    unsigned char res[16];
```
Comment (dgarske) [inline wolfcrypt/test/test.c:30561]: Should we have an enum/define for the SIP hash size 16 or is that variable?

**Review by dgarske — CHANGES_REQUESTED** [review]
In this case adding a new .c file for SIP hash makes sense, but did you consider putting it into an existing file? You'll need to add this .c to the CMake and VS projects also.

File: wolfcrypt/test/test.c (hunk 7 above)
Comment (SparkiDev) [reply wolfcrypt/test/test.c:30561]: There are two different calculations for different sizes and you are expected to get the output as 8 or 16.
So, it is both an identifier and a variable.
I'll add defines.

**Review by dgarske — CHANGES_REQUESTED** [review]
Please add the new .c to cmake and vs projects. Thanks

File: wolfssl/wolfcrypt/siphash.h [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+#define WOLF_CRYPT_SIPHASH_H
+
+#include <wolfssl/wolfcrypt/types.h>
+
+#if defined(WOLFSSL_SIPHASH)
+
+/* DESCRIPTION
+ *
+ * SipHash is a PseudoRandom Function (PRF) that can be used with small
+ * messagers (less than 256 bytes).
```
Comment (dgarske) [inline wolfssl/wolfcrypt/siphash.h:33]: Typo? ‘messages’

**Review by dgarske — APPROVED** [review]

**Pull request merged**